### PR TITLE
[Enhancement] Materialized views support refresh granularity splits

### LINKED ISSUE #12925: Optimize the refresh logic of materialized views
## Enhancement
If the content of a refresh is relatively large, the materialized view may not be refreshed successfully, so we need to temporarily support split refresh at partition granularity when the storage layer is not well done.

At the same time, for the large and wide table scenario of materialized views, how to determine the refresh is also a problem.

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #12925

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 26 / 106 (24.53%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|0|10|00.00%|[213, 214, 216, 217, 218, 219, 220, 221, 223, 225]|
|:large_blue_circle:|com/starrocks/scheduler/ExecuteOption.java|0|1|00.00%|[20]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|7|61|11.48%|[166, 176, 177, 178, 180, 181, 182, 183, 184, 185, 186, 188, 189, 190, 191, 194, 195, 196, 197, 198, 199, 200, 201, 202, 204, 205, 206, 209, 210, 211, 212, 215, 216, 217, 218, 219, 220, 222, 227, 228, 229, 230, 231, 232, 233, 234, 236, 237, 238, 239, 241, 242, 243, 680]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|5|20.00%|[3316, 3317, 3318, 3319]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|3|9|33.33%|[43, 47, 48, 51, 55, 56]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|6|50.00%|[631, 632, 633]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|8|10|80.00%|[225, 226]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|1|1|100.00%|[]|

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 69 / 109 (63.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|0|10|00.00%|[213, 214, 216, 217, 218, 219, 220, 221, 223, 225]|
|:large_blue_circle:|com/starrocks/scheduler/ExecuteOption.java|0|1|00.00%|[20]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|5|20.00%|[3316, 3317, 3318, 3319]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|6|50.00%|[631, 632, 633]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|42|64|65.62%|[175, 187, 227, 228, 229, 231, 236, 237, 238, 239, 240, 241, 242, 243, 245, 246, 247, 248, 250, 251, 252, 689]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|1|1|100.00%|[]|

**Comment by Astralidea** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/common/util/PropertyAnalyzer.java [hunk 1]
```diff
@@ -210,6 +211,22 @@ public static int analyzePartitionTimeToLive(Map<String, String> properties) thr
...
+    public static int analyzePartitionRefreshNumber(Map<String, String> properties) throws AnalysisException {
+        int partitionRefreshNumber = -1;
+        if (properties != null && properties.containsKey(PROPERTIES_PARTITION_REFRESH_NUMBER)) {
+            try {
+                partitionRefreshNumber = Integer.parseInt(properties.get(PROPERTIES_PARTITION_REFRESH_NUMBER));
+            } catch (NumberFormatException e) {
+                throw new AnalysisException("Partition Refresh Number: " + e.getMessage());
+            }
+            if (partitionRefreshNumber <= 0) {
+                throw new AnalysisException("Partition Refresh Number should larger than 0.");
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/common/util/PropertyAnalyzer.java:223]: how to unset refresh number?
what if the user do not specify the refresh number, what is the default logic?

File: fe/fe-core/src/main/java/com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java [hunk 2]
```diff
@@ -152,6 +170,86 @@ public void processTaskRun(TaskRunContext context) throws Exception {
...
         // insert execute successfully, update the meta of materialized view according to ExecPlan
         updateMeta(execPlan);
+
+        if (mvContext.hasNextBatchPartition()) {
+            generateNextTaskRun();
+        }
+    }
+
+    @VisibleForTesting
+    public void filterPartitionByRefreshNumber(Set<String> partitionsToRefresh, MaterializedView materializedView) {
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java:180]: have you test if one source base table partition has two target mv partitions can run successfully?

File: fe/fe-core/src/main/java/com/starrocks/common/util/PropertyAnalyzer.java (hunk 1 above)
Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/common/util/PropertyAnalyzer.java:223]: This feature will do in Support alter materialized view properties or later PR

File: fe/fe-core/src/main/java/com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java (hunk 2 above)
Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java:180]: In fact, this configuration has nothing to do with the source table, only the number of partitions of the materialized view table.

**Review by gengjun-git — DISMISSED** [review]

**Review by Youngwb — APPROVED** [review]

**Review by Youngwb — DISMISSED** [review]

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 69 / 113 (61.06%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|0|1|00.00%|[277]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|0|12|00.00%|[202, 210, 218, 219, 221, 222, 223, 224, 225, 226, 228, 230]|
|:large_blue_circle:|com/starrocks/scheduler/ExecuteOption.java|0|1|00.00%|[20]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|5|20.00%|[3320, 3321, 3322, 3323]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|6|50.00%|[638, 639, 640]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|42|64|65.62%|[175, 187, 227, 228, 229, 231, 236, 237, 238, 239, 240, 241, 242, 243, 245, 246, 247, 248, 250, 251, 252, 689]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|10|11|90.91%|[145]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|1|1|100.00%|[]|

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Review by gengjun-git — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 69 / 113 (61.06%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|0|1|00.00%|[277]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|0|12|00.00%|[202, 210, 218, 219, 221, 222, 223, 224, 225, 226, 228, 230]|
|:large_blue_circle:|com/starrocks/scheduler/ExecuteOption.java|0|1|00.00%|[20]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|5|20.00%|[3320, 3321, 3322, 3323]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|6|50.00%|[638, 639, 640]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|42|64|65.62%|[175, 187, 227, 228, 229, 231, 236, 237, 238, 239, 240, 241, 242, 243, 245, 246, 247, 248, 250, 251, 252, 689]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|10|11|90.91%|[145]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 69 / 113 (61.06%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|0|1|00.00%|[277]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|0|12|00.00%|[202, 210, 218, 219, 221, 222, 223, 224, 225, 226, 228, 230]|
|:large_blue_circle:|com/starrocks/scheduler/ExecuteOption.java|0|1|00.00%|[20]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|5|20.00%|[3320, 3321, 3322, 3323]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|6|50.00%|[638, 639, 640]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|42|64|65.62%|[175, 187, 227, 228, 229, 231, 236, 237, 238, 239, 240, 241, 242, 243, 245, 246, 247, 248, 250, 251, 252, 689]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|10|11|90.91%|[145]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|1|1|100.00%|[]|

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 69 / 113 (61.06%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|0|1|00.00%|[277]|
|:large_blue_circle:|com/starrocks/common/util/PropertyAnalyzer.java|0|12|00.00%|[202, 210, 218, 219, 221, 222, 223, 224, 225, 226, 228, 230]|
|:large_blue_circle:|com/starrocks/scheduler/ExecuteOption.java|0|1|00.00%|[20]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|1|5|20.00%|[3320, 3321, 3322, 3323]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|6|50.00%|[638, 639, 640]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|42|64|65.62%|[175, 187, 227, 228, 229, 231, 236, 237, 238, 239, 240, 241, 242, 243, 245, 246, 247, 248, 250, 251, 252, 689]|
|:large_blue_circle:|com/starrocks/catalog/TableProperty.java|10|11|90.91%|[145]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/clone/DynamicPartitionScheduler.java|1|1|100.00%|[]|

**Pull request merged**